src/main/java/dev/morling/onebrc/CalculateAverage_kaufco.java

```diff
@@ -0,0 +1,643 @@
+
+        // Get temperature. Read 4 chars at once. Possible cases:
+        // - A) "9.9\n" -> positive; digits in tempChars bytes #0, #2; next line at endOfs + 5
+        // - B) "99.9" "\n" -> positive; digits in tempChars bytes #0, #1, #3; next line at endOfs + 6
+        // - C) "-9.9" "\n" -> negative; digits in tempChars bytes #1, #3; next line at endOfs + 6
+        // - D) "-99." "9\n" -> negative; digits in tempChars bytes #1, #2 + fetch next needed; next line at endOfs + 7
+
+        int tempChars = buf.getInt(endOfs + 1);
+        int temperature;
+
+        int nameEnd = endOfs;
+        boolean isNeg = (tempChars & 0xff) == 0x2d;
+        if ((tempChars & 0xff0000) == 0x2e0000) {
+            if (isNeg) {
+                // C
+                temperature = 1000 - ((tempChars >> 8) & 0xf) * 10 - ((tempChars >> 24) & 0xf);
+                endOfs += 6;
+            }
+            else {
+                // B
+                temperature = 1000 + (tempChars & 0xf) * 100 + ((tempChars >> 8) & 0xf) * 10 + ((tempChars >> 24) & 0xf);
+                endOfs += 6;
+            }
+        }
+        else {
+            if (isNeg) {
+                // D
+                temperature = 1000 - ((tempChars >> 8) & 0xf) * 100 - ((tempChars >> 16) & 0xf) * 10 - (buf.get(endOfs + 5) & 0xf);
+                endOfs += 7;
+            }
+            else {
+                // A
+                temperature = 1000 + (tempChars & 0xf) * 10 + ((tempChars >> 16) & 0xf);
+                endOfs += 5;
+            }
+        }
+
+        recordSample(buf, startOfs, nameEnd, weatherData, stationNames, hash, temperature);
+        return endOfs;
+    }
+
+    /**
+     * Records a sample for a weather station.
+     *
+     * @param buf input file buffer
+     * @param nameStart name start byte offset in the buffer
+     * @param nameEnd name end byte offset in the buffer
+     * @param weatherData weather data hash table (see {@link AggregationResult#weatherData} for detailed format).
+     * @param stationNames buffer of weather station names (see {@link AggregationResult#stationNames} for detailed format).
+     * @param hash hash code for station name
+     * @param temperature temperature for the current sample
+     */
+    private static void recordSample(
+                                     ByteBuffer buf,
+                                     int nameStart,
+                                     int nameEnd,
+                                     ByteBuffer weatherData,
+                                     ByteBuffer stationNames,
+                                     int hash,
+                                     int temperature) {
+        int ofs = (hash & WEATHER_DATA_INDEX_MASK) << 4;
+        long lo = weatherData.getLong(ofs);
+        long hi = weatherData.getLong(ofs + 8);
+
+        int prevHash = (int) (lo >> 32);
+        if ((prevHash != hash) && (prevHash != 0))
+            throw new HashCollisionException();
```

Comment:
So what happens if there is a collision? Is there some means of open addressing fallback or similar? Can you point me to it? Note that the program should work for any valid keyset, so collisions must be handled somehow. To that end, when I test with 10K keys (see _create_measurements3.sh_), the program doesn't seem to finish (or it takes a very long time). While that's not the official evaluation, I'd like to make sure at least for the top entries that this keyset also can be processed, so as to make sure no corners are cut. Could you take a look? Thx.

Reply:
Yes: it falls back to the baseline implementation (`CalculateAverage_baseline`), so it's normal that it is getting slow. These are the relevant lines:

692 to 609:
```
        public void run() {
            try {
                processFile();
            }
            catch (IOException | HashCollisionException e) {
                abortRun = true;
            }
        }
```

and 202..206:
```
        if (abortRun) {
            // In case of hash collision: slooow fallback
            CalculateAverage_baseline.main(new String[0]);
            return;
        }
```

My original intention was to resort to a multi-set hash table implementation in the fallback case (which would probably have been only a few percent slower than the the single-set hash table implementation), but then I thought I don't need to since its not relevant for the competition result.

I think calling into the baseline class is OK? If not, I could just copy the code of it into my class, but that wouldn't change anything at the implementation, just code duplication.

Reply:
Each entry should be self-contained, i.e. calling out to the baseline should not be done. Would be nice to have a fallback which is reasonably fast. Keep in mind the goal is not to solely focus on the example data set but implement something which works in general for all allowed data sets.

Reply:
Fixed, see latest commit. I tested it with 100m entries @10k entries:
- execution time: 1.63s on my machine
- validated against output from merykitty